src/Adapter/MSTest.TestAdapter/Execution/TestClassInfo.cs

```diff
@@ -458,50 +483,80 @@ internal void ExecuteClassCleanup()
             }
             catch (Exception exception)
             {
-                var realException = exception.GetRealException();
-                ClassCleanupException = realException;
-
-                // special case AssertFailedException to trim off part of the stack trace
-                string errorMessage = realException is AssertFailedException or AssertInconclusiveException
-                    ? realException.Message
-                    : realException.GetFormattedExceptionMessage();
-
-                var exceptionStackTraceInfo = realException.TryGetStackTraceInformation();
-
-                var testFailedException = new TestFailedException(
-                    ObjectModelUnitTestOutcome.Failed,
-                    string.Format(
-                        CultureInfo.CurrentCulture,
-                        Resource.UTA_ClassCleanupMethodWasUnsuccesful,
-                        classCleanupMethod!.DeclaringType!.Name,
-                        classCleanupMethod.Name,
-                        errorMessage,
-                        exceptionStackTraceInfo?.ErrorStackTrace),
-                    exceptionStackTraceInfo,
-                    realException);
-                ClassCleanupException = testFailedException;
-
-                throw testFailedException;
+                ClassCleanupException = exception;
             }
         }
+
+        // If ClassCleanup was successful, then don't do anything
+        if (ClassCleanupException == null)
+        {
+            // We didn't but it in a finally function above to not change the old behavior as it was only in the the try
+            IsClassCleanupExecuted = true;
+            return;
+        }
+
+        if (ClassCleanupException is TestFailedException)
+        {
+            throw ClassCleanupException;
+        }
+
+        var realException = ClassCleanupException.GetRealException();
+
+        // special case AssertFailedException to trim off part of the stack trace
+        string errorMessage = realException is AssertFailedException or AssertInconclusiveException
+            ? realException.Message
+            : realException.GetFormattedExceptionMessage();
+
+        var exceptionStackTraceInfo = realException.TryGetStackTraceInformation();
+
+        var testFailedException = new TestFailedException(
+            ObjectModelUnitTestOutcome.Failed,
+            string.Format(
+                CultureInfo.CurrentCulture,
+                Resource.UTA_ClassCleanupMethodWasUnsuccesful,
+                classCleanupMethod!.DeclaringType!.Name,
+                classCleanupMethod.Name,
+                errorMessage,
+                exceptionStackTraceInfo?.ErrorStackTrace),
+            exceptionStackTraceInfo,
+            realException);
+        ClassCleanupException = testFailedException;
+
+        throw testFailedException;
+    }
+
+    private TestFailedException? InvokeCleanupMethod(MethodInfo? methodInfo)
+    {
+        if (methodInfo is null)
```

Comment:
do we expect a null method? Isn't confusing...we return null in 2 case but the semantic is completely different one is a "wrong" parameter the other one is and invocation without errrors.

I would throw if null or?